### ISSUE TITLE
TS2488: Generic type parameter not inferred from JSX props, falls back to default

### ISSUE THREAD

**Issue by theoludwig** [microsoft/typescript-go#2797]
tsgo reports a false positive `TS2488` when using a generic JSX component where a type parameter should be inferred from one prop (`selector`) and consumed in another (`children`). 

tsgo fails to infer the type parameter from the JSX prop and falls back to the default, causing a spurious error. The equivalent non-JSX function call works correctly.

`tsc` (TypeScript 5.9.3) reports no error on the same code.

### Minimal Reproduction

https://github.com/theoludwig/oxlint-tsgolint-issue-135

```tsx
interface State {
  value: boolean
}

declare const Subscribe: <TSelected = State>(props: {
  selector?: (state: State) => TSelected
  children: (state: TSelected) => void
}) => React.ReactNode

const _result = (
  <Subscribe
    selector={(state) => {
      return [state.value]
    }}
  >
    {([value = false]) => {
      console.log(value)
    }}
  </Subscribe>
)
```

### Expected behavior

`TSelected` should be inferred as `boolean[]` from the `selector` return type (same as `tsc`). No error should be reported.

### Actual behavior

tsgo does not infer `TSelected` from the `selector` prop and falls back to the default (`State`). This causes:

```
TS2488: Type 'State' must have a '[Symbol.iterator]()' method that returns an iterator.
```

on the array destructuring pattern `[value = false]` in `children`.

### Key observation: JSX-specific

The **same pattern** as a regular function call works correctly in tsgo, no error:

```ts
Subscribe({
  selector: (state) => {
    return [state.value]
  },
  children: ([value = false]) => {
    console.log(value)
  },
})
```

The bug only manifests when the generic function is used as a JSX component with `children` passed as JSX children (between opening and closing tags) rather than as a direct prop.

### Environment

- Detected via `oxlint 1.48.0` with `--type-check --type-aware` (which uses tsgo via `oxlint-tsgolint 0.14.0`)
- `tsc` (TypeScript 5.9.3): no error
- This issue was not present with `oxlint-tsgolint 0.12.2` and appeared after `0.13.0` which updated the typescript-go submodule

### Real-world impact

This affects all uses of `@tanstack/react-form`'s `form.Subscribe` component, which uses exactly this pattern — a `selector` prop that transforms `FormState` into a derived value, and `children` that receives the selected value.

**Comment by jakebailey** [comment]
Can you please check against TS 6.0, not 5.9?

**Comment by theoludwig** [comment]
> Can you please check against TS 6.0, not 5.9?

Sure.
I tried the repro with both:
- TypeScript v6 beta (`typescript@6.0.0-beta`), `tsc --noEmit` => no errors reported (expected behavior)
- TypeScript v7 (`@typescript/native-preview@7.0.0-dev.20260216.1`) `tsgo --noEmit` => reports error TS2488

https://github.com/theoludwig/oxlint-tsgolint-issue-135/commit/35831df5aab8710c71d4094a2590a55b3541c208

The output is as follows:
```ts
➜ node --run typescript6

➜ node --run typescript7
src/main.tsx:15:5 - error TS2322: Type '(state: State) => boolean[]' is not assignable to type '(state: State) => State'.
  Property 'value' is missing in type 'boolean[]' but required in type 'State'.

15     selector={(state) => {
       ~~~~~~~~

  src/main.tsx:5:3 - 'value' is declared here.
    5   value: boolean
        ~~~~~

  src/main.tsx:9:3 - The expected type comes from property 'selector' which is declared here on type 'IntrinsicAttributes & { selector?: ((state: State) => State) | undefined; children: (state: State) => void; }'
    9   selector?: (state: State) => TSelected
        ~~~~~~~~

src/main.tsx:19:7 - error TS2488: Type 'State' must have a '[Symbol.iterator]()' method that returns an iterator.

19     {([value = false]) => {
         ~~~~~~~~~~~~~~~


Found 2 errors in the same file, starting at: src/main.tsx:15
```

It seems to be a regression in TypeScript v7, because a few days ago (I haven't bisected the exact version/commit), it worked (works with the TypeScript v7 version used by https://github.com/oxc-project/tsgolint v0.12.2).

It hasn't worked since https://github.com/oxc-project/tsgolint/pull/680 (oxc-project/tsgolint v0.13.0 and above)

Should be related to one of those commits:
https://github.com/microsoft/typescript-go/compare/5597f4c8ecf49b4344b8280ab74c01c46e85d78a...ee33e62c917c88537c17ec426a28ad483379f5a8

(to be clear, it doesn't seem to be an issue with oxc/tsgolint, as the issue can be reproduced independently as shown above, however oxc/tsgolint is how I discovered the issue)

**Comment by theoludwig** [comment]
This PR seems to have changed the behavior related to `[Symbol.iterator]()`: https://github.com/microsoft/typescript-go/pull/2737
But it has been implemented in TypeScript v6 (from what I understand): https://github.com/microsoft/typescript/issues/62661
so I don't know if it's related or not, but hopefully it helps! :smile:

**Comment by theoludwig** [comment]
Ah, I found the problematic commit: https://github.com/microsoft/typescript-go/pull/2773

I tried the following (inside the repro: https://github.com/theoludwig/oxlint-tsgolint-issue-135)
```sh
git clone --recurse-submodules https://github.com/microsoft/typescript-go.git
cd typescript-go
npm clean-install
npx hereby build

# go back to repro
cd ..
./typescript-go/built/local/tsgo --noEmit
# same errors as @typescript/native-preview@7.0.0-dev.20260216.1
```

However, if I revert this PR: https://github.com/microsoft/typescript-go/pull/2773
```sh
cd typescript-go
git revert 80401b21fc328ccbc635ca48bf55bad30aeac128
npx hereby build

# go back to repro
cd ..
./typescript-go/built/local/tsgo --noEmit
# no errors, expected behavior, works correctly!
```
